File: CS/GridBeforePrint/MyGridViewPrintInfo.cs
using DevExpress.Utils;
using DevExpress.Utils.Text;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Views.Printing;
using DevExpress.XtraPrinting;
using System;
using System.Drawing;

namespace GridBeforePrint
{
    public class MyGridViewPrintInfo : GridViewPrintInfo
    {
        public MyGridViewPrintInfo(PrintInfoArgs args) : base(args) { }

        public override void PrintHeader(DevExpress.XtraPrinting.IBrickGraphics graph)
        {
            if (!View.OptionsPrint.PrintHeader) return;
            Point indent = new Point(Indent, HeaderY);
            Rectangle r = Rectangle.Empty;
            bool usePrintStyles = View.OptionsPrint.UsePrintStyles;
            SetDefaultBrickStyle(graph, Bricks["HeaderPanel"]);
            foreach (PrintColumnInfo col in Columns)
            {
                if (!usePrintStyles)
                {
                    AppearanceObject temp = new AppearanceObject();
                    AppearanceHelper.Combine(temp, new AppearanceObject[] { col.Column.AppearanceHeader, View.Appearance.HeaderPanel, AppearancePrint.HeaderPanel });
                    SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1));
                }
                r = col.Bounds;
                r.Offset(indent);
                string caption = GetTextCaptionForPrinting(col.Column);
                if (!col.Column.OptionsColumn.ShowCaption) caption = string.Empty;
                ITextBrick itb = DrawTextBrick(graph, caption, r);

                (View as MyGridView).RaiseHeaderPrintEvent(this, new HeaderPrintEventArgs(itb, col));

                if (caption.Contains(Environment.NewLine)) itb.Style.StringFormat = BrickStringFormat.Create(itb.Style.TextAlignment, true);
                if (AppearancePrint.HeaderPanel.TextOptions.WordWrap == WordWrap.NoWrap && View.OptionsPrint.UsePrintStyles)
                {
                    using (Graphics g = this.View.GridControl.CreateGraphics())
                    {
                        SizeF s = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value);
                        if (s.Width + 5 >= r.Width)
                        {
                            itb.Text = "";
                            itb.TextValue = "";
                        }
                    }
                }
            }
        }

        protected string GetTextCaptionForPrinting(GridColumn column)
        {
            if (View == null) return column.Caption;
            var gview = View.OptionsPrint as GridOptionsPrint;
            if (gview != null && !gview.AllowMultilineHeaders) return column.GetTextCaption();
            return (View as MyGridView).GoAndGetNonFormattedCaption(column.GetCaption());
        }

        protected override void PrintRowCell(int rowHandle, GridCellInfo cell, Rectangle r)
        {
            View.OptionsPrint.PrintPreview = true;
            string displayText = (View as MyGridView).GoAndGetRowCellDisplayTextCore(rowHandle, cell.Column, cell.ViewInfo, cell.CellValue, false);
            MyGridViewInfo myGridViewInfo = new MyGridViewInfo(View as MyGridView);
            myGridViewInfo.GoAndUpdateCellAppearanceCore(cell);
            if (cell.ViewInfo.AllowHtmlString)
            {
                displayText = StringPainter.Default.RemoveFormat(displayText, true);
            }
            HorzAlignment horzAlignment = (View as MyGridView).GetHorzAlignment(rowHandle, cell.Column, cell.Appearance.HAlignment);
            PrintCellHelperInfo info = new PrintCellHelperInfo(new Point(cell.Column == null ? -1 : cell.Column.AbsoluteIndex, rowHandle),
                LineColor,
                PS,
                cell.CellValue,
                cell.Appearance,
                displayText,
                r,
                Graph,
                horzAlignment,
                View.OptionsPrint.PrintHorzLines,
                View.OptionsPrint.PrintVertLines,
                cell.ColumnInfo.Column.DisplayFormat.FormatString,
                CalcBrickBordersEX(),
                new PaddingInfo(2, GraphicsDpi.Pixel)
                );

            IVisualBrick brick = cell.Editor.GetBrick(info);

            MyGridView view = this.View as MyGridView;
            view.RaiseSamplePrintEvent(this, new SamplePrintEventArgs(cell.RowHandle, cell.Column, brick, false));
            
            if (AllowProcessMergedInfo)
            {
                brick.Rect = r;
                UpdateMergedStatus(cell, (VisualBrick)brick);
            }
            Graph.DrawBrick(brick, r);
        }

        DevExpress.XtraPrinting.BorderSide CalcBrickBordersEX()
        {
            BorderSide border = BorderSide.None;
            if (View.OptionsPrint.PrintHorzLines) border |= BorderSide.Top | BorderSide.Bottom;
            if (View.OptionsPrint.PrintVertLines) border |= BorderSide.Left | BorderSide.Right;
            return border;
        }

    }
}

File: VB/GridBeforePrint/MyGridViewPrintInfo.vb
Imports DevExpress.Utils
Imports DevExpress.Utils.Text
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Views.Printing
Imports DevExpress.XtraPrinting
Imports System
Imports System.Drawing

Namespace GridBeforePrint
    Public Class MyGridViewPrintInfo
        Inherits GridViewPrintInfo

        Public Sub New(ByVal args As PrintInfoArgs)
            MyBase.New(args)
        End Sub

        Public Overrides Sub PrintHeader(ByVal graph As BrickGraphics)
            If Not View.OptionsPrint.PrintHeader Then
                Return
            End If
            Dim indent As New Point(Me.Indent, HeaderY)
            Dim r As Rectangle = Rectangle.Empty
            Dim usePrintStyles As Boolean = View.OptionsPrint.UsePrintStyles
            SetDefaultBrickStyle(graph, Bricks("HeaderPanel"))
            For Each col As PrintColumnInfo In Columns
                If Not usePrintStyles Then
                    Dim temp As New AppearanceObject()
                    AppearanceHelper.Combine(temp, New AppearanceObject() {col.Column.AppearanceHeader, View.Appearance.HeaderPanel, AppearancePrint.HeaderPanel})
                    SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1))
                End If
                r = col.Bounds
                r.Offset(indent)
                Dim caption As String = GetTextCaptionForPrinting(col.Column)
                If Not col.Column.OptionsColumn.ShowCaption Then
                    caption = String.Empty
                End If
                Dim itb As ITextBrick = DrawTextBrick(graph, caption, r)

                TryCast(View, MyGridView).RaiseHeaderPrintEvent(Me, New HeaderPrintEventArgs(itb, col))

                If caption.Contains(Environment.NewLine) Then
                    itb.Style.StringFormat = BrickStringFormat.Create(itb.Style.TextAlignment, True)
                End If
                If AppearancePrint.HeaderPanel.TextOptions.WordWrap = WordWrap.NoWrap AndAlso View.OptionsPrint.UsePrintStyles Then
                    Using g As Graphics = Me.View.GridControl.CreateGraphics()
                        Dim s As SizeF = g.MeasureString(itb.Text, itb.Font, 1000, itb.StringFormat.Value)
                        If s.Width + 5 >= r.Width Then
                            itb.Text = ""
                            itb.TextValue = ""
                        End If
                    End Using
                End If
            Next col
        End Sub

        Protected Function GetTextCaptionForPrinting(ByVal column As GridColumn) As String
            If View Is Nothing Then
                Return column.Caption
            End If
            Dim gview = TryCast(View.OptionsPrint, GridOptionsPrint)
            If gview IsNot Nothing AndAlso (Not gview.AllowMultilineHeaders) Then
                Return column.GetTextCaption()
            End If
            Return (TryCast(View, MyGridView)).GoAndGetNonFormattedCaption(column.GetCaption())
        End Function

        Protected Overrides Sub PrintRowCell(ByVal rowHandle As Integer, ByVal cell As GridCellInfo, ByVal r As Rectangle)
            Me.View.OptionsPrint.PrintPreview = True
            Dim displayText As String = (TryCast(Me.View, MyGridView)).GoAndGetRowCellDisplayTextCore(rowHandle, cell.Column, cell.ViewInfo, cell.CellValue, False)
            Dim myGridViewInfo As New MyGridViewInfo(TryCast(Me.View, MyGridView))
            myGridViewInfo.GoAndUpdateCellAppearanceCore(cell)
            If cell.ViewInfo.AllowHtmlString Then
                displayText = StringPainter.Default.RemoveFormat(displayText, True)
            End If
            Dim horzAlignment As HorzAlignment = (TryCast(Me.View, MyGridView)).GetHorzAlignment(rowHandle, cell.Column, cell.Appearance.HAlignment)
            Dim info As New PrintCellHelperInfo(New Point(If(cell.Column Is Nothing, -1, cell.Column.AbsoluteIndex), rowHandle), LineColor, PrintingSystemBase, cell.CellValue, cell.Appearance, displayText, r, Graph, horzAlignment, Me.View.OptionsPrint.PrintHorzLines, Me.View.OptionsPrint.PrintVertLines, cell.ColumnInfo.Column.DisplayFormat.FormatString, CalcBrickBordersEX(), PaddingInfo.Empty)

            Dim brick As VisualBrick = cell.Editor.GetBrick(info)

            Dim view As MyGridView = TryCast(Me.View, MyGridView)
            view.RaiseSamplePrintEvent(Me, New SamplePrintEventArgs(cell.RowHandle, cell.Column, brick, False))

            If AllowProcessMergedInfo Then
                brick.Rect = r
                UpdateMergedStatus(cell, DirectCast(brick, VisualBrick))
            End If
            Graph.DrawBrick(brick, r)
        End Sub

        Private Function CalcBrickBordersEX() As DevExpress.XtraPrinting.BorderSide
            Dim border As BorderSide = BorderSide.None
            If View.OptionsPrint.PrintHorzLines Then
                border = border Or BorderSide.Top Or BorderSide.Bottom
            End If
            If View.OptionsPrint.PrintVertLines Then
                border = border Or BorderSide.Left Or BorderSide.Right
            End If
            Return border
        End Function

    End Class
End Namespace
